### PR TITLE
Features/provisioning sdk rbac e2 e tests

### PR DESCRIPTION
**Pull request by dylanbulfinMS**
## Checklist
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/main/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [ ] This pull-request is submitted against the `main` branch.

### CODE REVIEW

File: e2e/test/config/TestConfiguration.Provisioning.cs [hunk 1]
```diff
@@ -38,6 +76,77 @@ public static X509Certificate2Collection GetGroupEnrollmentChain()
...
+
+                if (!string.IsNullOrWhiteSpace(policyName))
+                {
+                    token += "&skn=" + policyName;
+                }
+
+                return token;
+            }
+
+            public class ConnectionStringParser
```
Comment (drwill-ms) [inline e2e/test/config/TestConfiguration.Provisioning.cs:109]: Prefer to put classes in their own file.

File: e2e/test/config/TestConfiguration.Provisioning.cs [hunk 2]
```diff
@@ -38,6 +76,77 @@ public static X509Certificate2Collection GetGroupEnrollmentChain()
...
+            {
+                public ConnectionStringParser(string connectionString)
+                {
+                    string[] parts = connectionString.Split(';');
```
Comment (drwill-ms) [inline e2e/test/config/TestConfiguration.Provisioning.cs:113]: Consider splitting on ; and =, and putting the kvps in a dictionary.

File: e2e/test/config/TestConfiguration.Provisioning.cs [hunk 3]
```diff
@@ -38,6 +76,77 @@ public static X509Certificate2Collection GetGroupEnrollmentChain()
...
+                                SharedAccessKeyName = part.Substring("SHAREDACCESSKEYNAME=".Length);
+                                break;
+
+                            default:
+                                throw new NotSupportedException("Unrecognized tag found in test ConnectionString.");
+                        }
+                    }
+                }
+
+                public string ProvisioningHostName { get; private set; }
```
Comment (drwill-ms) [inline e2e/test/config/TestConfiguration.Provisioning.cs:142]: properties above methods (ordering in a class)

File: e2e/test/config/TestConfiguration.Provisioning.cs [hunk 4]
```diff
@@ -38,6 +76,77 @@ public static X509Certificate2Collection GetGroupEnrollmentChain()
...
+
+                public string DeviceID { get; private set; }
```
Comment (drwill-ms) [inline e2e/test/config/TestConfiguration.Provisioning.cs:144]:
```suggestion
                public string DeviceId { get; private set; }
```

File: e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs [hunk 5]
```diff
@@ -110,6 +111,45 @@ public async Task ProvisioningServiceClient_GetEnrollmentGroupAttestation_Symmet
...
         }
 
+        [LoggedTestMethod]
+        public async Task ProvisioningServiceClient_TokenCredentialAuth_Success()
+        {
+            using var provisioningServiceClient = ProvisioningServiceClient.Create(
+                TestConfiguration.Provisioning.GetProvisioningHostName(),
+                TestConfiguration.Provisioning.GetClientSecretCredential());
+
+            IndividualEnrollment individualEnrollment = await CreateIndividualEnrollment(provisioningServiceClient, AttestationMechanismType.SymmetricKey, null, AllocationPolicy.Static, null, null, null);
```
Comment (drwill-ms) [inline e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:121]: When lines get this long, we like to break them up. For this one, parameters can each get their own line. Also, in product code and tests (not samples) we append `.ConfigureAwait(false)` to async calls.

```suggestion
            IndividualEnrollment individualEnrollment = await CreateIndividualEnrollment(
                    provisioningServiceClient,
                    AttestationMechanismType.SymmetricKey,
                    null,
                    AllocationPolicy.Static,
                    null,
                    null,
                    null)
                .ConfigureAwait(false);
```

File: e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs [hunk 6]
```diff
@@ -110,6 +111,45 @@ public async Task ProvisioningServiceClient_GetEnrollmentGroupAttestation_Symmet
...
+
+            //act
+            IndividualEnrollment individualEnrollmentResult = await provisioningServiceClient.CreateOrUpdateIndividualEnrollmentAsync(individualEnrollment);
+
+            //assert
+            Assert.IsNotNull(individualEnrollmentResult);
```
Comment (drwill-ms) [inline e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:127]: Prefer FluentAssertions.

```suggestion
            individualEnrollmentResult.Should().NotBeNull();
```

**Review by drwill-ms — CHANGES_REQUESTED** [review]
Inline comments.

File: e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs [hunk 7]
```diff
@@ -110,6 +111,45 @@ public async Task ProvisioningServiceClient_GetEnrollmentGroupAttestation_Symmet
...
+
+        [LoggedTestMethod]
+        public async Task ProvisioningServiceClient_AzureSasCredentialAuth_Success()
+        {
+            string signature = TestConfiguration.Provisioning.GetProvisioningSharedAccessSignature(TimeSpan.FromHours(1));
+            using var provisioningServiceClient = ProvisioningServiceClient.Create(
+                TestConfiguration.Provisioning.GetProvisioningHostName(),
+                new AzureSasCredential(signature));
+
+            IndividualEnrollment individualEnrollment = await CreateIndividualEnrollment(provisioningServiceClient, AttestationMechanismType.SymmetricKey, null, AllocationPolicy.Static, null, null, null);
```
Comment (azabbasi) [inline e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:141]:
```suggestion
            IndividualEnrollment individualEnrollment = await CreateIndividualEnrollment(
                    provisioningServiceClient, 
                    AttestationMechanismType.SymmetricKey, 
                    null, 
                    AllocationPolicy.Static, 
                    null, 
                    null, 
                    null);
```

File: e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs [hunk 8]
```diff
@@ -110,6 +111,45 @@ public async Task ProvisioningServiceClient_GetEnrollmentGroupAttestation_Symmet
...
+
+            //act
+            IndividualEnrollment individualEnrollmentResult = await provisioningServiceClient.CreateOrUpdateIndividualEnrollmentAsync(individualEnrollment);
+
+            //assert
+            Assert.IsNotNull(individualEnrollmentResult);
```
Comment (azabbasi) [inline e2e/test/provisioning/ProvisioningServiceClientE2ETests.cs:147]: Please use fluent assertion instead.

File: e2e/test/provisioning/ConnectionStringParser.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
+﻿using System;
+
+namespace Microsoft.Azure.Devices.E2ETests.provisioning
```
Comment (jamdavi) [inline e2e/test/provisioning/ConnectionStringParser.cs:3]: Casing on the namespace.

```suggestion
namespace Microsoft.Azure.Devices.E2ETests.Provisioning
```

File: e2e/test/config/TestConfiguration.Provisioning.cs [hunk 10]
```diff
@@ -1,8 +1,19 @@
...
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+using System.Globalization;
+using System.Net;
+using System.Security.Cryptography;
 using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using Microsoft.Azure.Devices.E2ETests.provisioning;
```
Comment (jamdavi) [inline e2e/test/config/TestConfiguration.Provisioning.cs:10]: Will need to be changed with other casing request.

```suggestion
using Microsoft.Azure.Devices.E2ETests.Provisioning;
```

**Review by drwill-ms — APPROVED** [review]

**Review by azabbasi — APPROVED** [review]

**Pull request merged**